lib/pbench/server/api/resources/compare_dataset.py

```diff
@@ -0,0 +1,145 @@
+            ApiSchema(
+                ApiMethod.GET,
+                OperationCode.READ,
+                query_schema=Schema(
+                    Parameter(
+                        "datasets", ParamType.LIST, string_list=",", required=True
+                    ),
+                ),
+                authorization=ApiAuthorizationType.NONE,
+            ),
+        )
+
+    def _get(
+        self, params: ApiParams, request: Request, context: ApiContext
+    ) -> Response:
+        """
+        This function is using Quisby to compare results into a form that supports visualization
+
+        Args:
+            params: includes the uri parameters, which provide the list of dataset.
+            request: Original incoming Request object
+            context: API context dictionary
+
+        Raises:
+            UnauthorizedAccess : The user isn't authorized for the requestedaccess.
+            APIAbort, reporting "NOT_FOUND" and "INTERNAL_SERVER_ERROR"
+            APIInternalError, reporting the failure message
+
+        GET /api/v1/compare?datasets=d1,d2,d3
+        """
+
+        datasets = params.query.get("datasets")
+        authorized_user: User = Auth.token_auth.current_user()
+
+        if not authorized_user:
+            raise APIAbort(
+                HTTPStatus.UNAUTHORIZED,
+                "User provided access_token is invalid or expired",
+            )
+
+        ds_list = [convert_dataset(ds, None) for ds in datasets]
+        cache_m = CacheManager(self.config, current_app.logger)
+
+        benchmark_check = []
+
+        stream_file = {}
+        for dataset in ds_list:
+            metadata = self._get_dataset_metadata(
+                dataset, ["dataset.metalog.pbench.script"]
+            )
+
+            benchmark = metadata["dataset.metalog.pbench.script"]
+            benchmark_check.append(benchmark)
+
+            # Validate if all the selected datasets is of same benchmark
+            if len(set(benchmark_check)) != 1:
+                raise APIAbort(
+                    HTTPStatus.BAD_REQUEST,
+                    "Requested datasets are not of same benchmark. It can't be compared",
+                )
+
+            # Validate if the user is authorized to access the selected datasets
+            if (
+                dataset.owner_id != str(authorized_user.id)
+                and dataset.access == Dataset.PRIVATE_ACCESS
+            ):
+                raise UnauthorizedAccess(
+                    authorized_user,
+                    OperationCode.READ,
+                    dataset.owner.username,
+                    dataset.access,
+                    HTTPStatus.FORBIDDEN,
+                )
+
+        for dataset in ds_list:
+            try:
+                tarball = cache_m.find_dataset(dataset.resource_id)
+            except TarballNotFound as e:
+                raise APIAbort(
+                    HTTPStatus.NOT_FOUND, f"No dataset with ID '{e.tarball}' found"
+                ) from e
```

Comment:
Since all the datasets have been validated already, this means we've got an inconsistency between our SQL and the cache manager, which is an internal server error. I think this should be `APIInternalError` with a message saying that the expected dataset ID is missing from the cache manager.